lib/resty/aes.lua

```diff
@@ -125,50 +125,75 @@ cipher = function (size, _cipher)
+
+local options = {
+    default = 0x01,
+    key_without_gen = 0x02
+    -- may add other things like padding mode; or use bit 'or/and' operation for combined option
+}
+_M.options = options
+
+function _M.new(self, key, salt, _cipher, _hash, hash_rounds, option)
     local encrypt_ctx = ffi_new(ctx_ptr_type)
     local decrypt_ctx = ffi_new(ctx_ptr_type)
     local _cipher = _cipher or cipher()
     local _hash = _hash or hash.md5
     local hash_rounds = hash_rounds or 1
+    local option = option or options.default
     local _cipherLength = _cipher.size/8
     local gen_key = ffi_new("unsigned char[?]",_cipherLength)
     local gen_iv = ffi_new("unsigned char[?]",_cipherLength)
 
-    if type(_hash) == "table" then
-        if not _hash.iv or #_hash.iv ~= 16 then
-          return nil, "bad iv"
-        end
+    if option == options.default then
+        if type(_hash) == "table" then
+            if not _hash.iv or #_hash.iv ~= 16 then
+              return nil, "bad iv"
+            end
+
+            if _hash.method then
+                local tmp_key = _hash.method(key)
 
-        if _hash.method then
-            local tmp_key = _hash.method(key)
+                if #tmp_key ~= _cipherLength then
+                    return nil, "bad key length"
+                end
 
-            if #tmp_key ~= _cipherLength then
+                ffi_copy(gen_key, tmp_key, _cipherLength)
+
+            elseif #key ~= _cipherLength then
                 return nil, "bad key length"
-            end
 
-            ffi_copy(gen_key, tmp_key, _cipherLength)
+            else
+                ffi_copy(gen_key, key, _cipherLength)
+            end
 
-        elseif #key ~= _cipherLength then
-            return nil, "bad key length"
+            ffi_copy(gen_iv, _hash.iv, 16)
 
         else
-            ffi_copy(gen_key, key, _cipherLength)
-        end
+            if salt and #salt ~= 8 then
+                return nil, "salt must be 8 characters or nil"
+            end
 
-        ffi_copy(gen_iv, _hash.iv, 16)
+            if C.EVP_BytesToKey(_cipher.method, _hash, salt, key, #key,
+                                hash_rounds, gen_key, gen_iv)
+                ~= _cipherLength
+            then
+                return nil
+            end
+        end
 
-    else
-        if salt and #salt ~= 8 then
-            return nil, "salt must be 8 characters or nil"
+    elseif option == options.key_without_gen then
+        -- use origin key
+        if not key then
+            return nil, "key is nil!"
         end
 
-        if C.EVP_BytesToKey(_cipher.method, _hash, salt, key, #key,
-                            hash_rounds, gen_key, gen_iv)
-            ~= _cipherLength
-        then
-            return nil
+        if #key > _cipherLength then
+            return nil, "key too long"
         end
+
+        -- note it's key padding not data padding, and it use zero padding only;
+        key = key..string.rep("\0", _cipherLength - #key)
```

Comment:
Please avoid this global variable `string`. We should avoid use of Lua globals on such Lua hot code paths. We should write

```
local str_rep = string.rep

function blah(...)
    str_rep(...)
end
```

This way we also save 2 table looks on hot code paths, one is the table lookup of the key "string" in the global environment table, and the other is the lookup of the key "rep" in the "string" table.